crates/sui-core/src/batch_bls_verifier.rs

```diff
@@ -1,37 +1,304 @@
+            |_| {
+                self.cache.cache_certs_verified(
+                    certs.into_iter().map(|c| c.certificate_digest()).collect(),
+                )
+            },
+        )
+    }
+
+    /// Verifies one cert asynchronously, in a batch.
+    pub async fn verify_cert(&self, cert: CertifiedTransaction) -> SuiResult<VerifiedCertificate> {
+        let cert_digest = cert.certificate_digest();
+        if self.cache.is_cert_verified(&cert_digest) {
+            return Ok(VerifiedCertificate::new_unchecked(cert));
+        }
+        self.verify_cert_skip_cache(cert)
+            .await
+            .tap_ok(|_| self.cache.cache_cert_verified(cert_digest))
+    }
+
+    /// exposed as a public method for the benchmarks
+    pub async fn verify_cert_skip_cache(
+        &self,
+        cert: CertifiedTransaction,
+    ) -> SuiResult<VerifiedCertificate> {
+        // this is the only innocent error we are likely to encounter - filter it before we poison
+        // a whole batch.
+        if cert.auth_sig().epoch != self.committee.epoch() {
+            return Err(SuiError::WrongEpoch {
+                expected_epoch: self.committee.epoch(),
+                actual_epoch: cert.auth_sig().epoch,
+            });
+        }
+
+        cert.verify_sender_signatures()?;
+        self.verify_cert_inner(cert).await
+    }
+
+    async fn verify_cert_inner(
+        &self,
+        cert: CertifiedTransaction,
+    ) -> SuiResult<VerifiedCertificate> {
+        let (tx, rx) = oneshot::channel();
+        pin_mut!(rx);
+
+        let prev_id = {
+            let mut queue = self.queue.lock();
+            queue.push(tx, cert);
+            if queue.len() == queue.capacity() {
+                self.metrics.full_batches.inc();
+                self.process_queue(queue);
+                // unwrap ok - process_queue will have sent the result already
+                return rx.try_recv().unwrap();
+            }
+            queue.id
+        };
+
+        if let Ok(res) = timeout(BATCH_TIMEOUT_MS, &mut rx).await {
+            // unwrap ok - tx cannot have been dropped without sending a result.
+            return res.unwrap();
+        }
+        self.metrics.timeouts.inc();
+
+        {
+            let queue = self.queue.lock();
+            // check if another thread took the queue while we were re-acquiring lock.
+            if prev_id == queue.id {
+                debug_assert_ne!(queue.len(), queue.capacity());
+                self.metrics.partial_batches.inc();
+                self.process_queue(queue);
+                // unwrap ok - process_queue will have sent the result already
+                return rx.try_recv().unwrap();
+            }
+        }
+
+        // unwrap ok - another thread took the queue while we were re-acquiring the lock and is
+        // guaranteed to process the queue immediately.
+        rx.await.unwrap()
```

Comment:
Is this method cancellation safe? Is it possible that task grabs the queue, then that particular task is dropped and queue is never processed (therefore this unwrap panics?).

Reply:
Its safe - the only await point is the timeout, and the locks are all sync so they can't be held across an await even if you wanted to. Therefore once a task has taken the queue it is guaranteed to finish.

Reply:
Ok, yes, I think you are right. Maybe mention this in the comment (that cancellation safety is achieved because this is a sync lock not held across await points)